src/transformers/models/propainter/modeling_propainter.py

```diff
@@ -0,0 +1,4630 @@
+
+    def forward(self, hidden_states, batch_size, output_size):
+        features_height = int(
+            (output_size[0] + 2 * self.padding[0] - (self.kernel_size[0] - 1) - 1) / self.stride[0] + 1
+        )
+        features_width = int(
+            (output_size[1] + 2 * self.padding[1] - (self.kernel_size[1] - 1) - 1) / self.stride[1] + 1
+        )
+
+        hidden_states = self.unfold(hidden_states)
+        hidden_states = hidden_states.permute(0, 2, 1)
+        hidden_states = self.embedding(hidden_states)
+        hidden_states = hidden_states.view(batch_size, -1, features_height, features_width, hidden_states.size(2))
+
+        return hidden_states
+
+
+class ProPainterSoftComp(nn.Module):
+    def __init__(self, config: ProPainterConfig):
+        super().__init__()
+        self.config = config
+        self.relu = nn.LeakyReLU(0.2, inplace=True)
+        output_features = reduce((lambda x, y: x * y), config.kernel_size) * config.num_channels
+        self.embedding = nn.Linear(config.hidden_size, output_features)
+        self.kernel_size = config.kernel_size
+        self.stride = config.stride
+        self.padding = config.padding_inpaint_generator
+        self.bias_conv = nn.Conv2d(
+            config.num_channels,
+            config.num_channels,
+            kernel_size=config.patch_size,
+            stride=1,
+            padding=config.padding,
+        )
+
+    def forward(self, hidden_state, timestep, output_size):
+        num_batch_, _, _, _, channel_ = hidden_state.shape
+        hidden_state = hidden_state.view(num_batch_, -1, channel_)
+        hidden_state = self.embedding(hidden_state)
+        batch_size, _, num_channels = hidden_state.size()
+        hidden_state = hidden_state.view(batch_size * timestep, -1, num_channels).permute(0, 2, 1)
+        hidden_state = F.fold(
+            hidden_state,
+            output_size=output_size,
+            kernel_size=self.kernel_size,
+            stride=self.stride,
+            padding=self.padding,
+        )
+        hidden_state = self.bias_conv(hidden_state)
+
+        return hidden_state
+
+
+def window_partition(input_feature, window_size, num_attention_heads):
+    """
+    Args:
+        input_feature: shape is (batch_size, timesteps, height, width, num_channels)
+        window_size (tuple[int]): window size
+    Returns:
+        windows: (batch_size, num_windows_h, num_windows_w, num_attention_heads, timesteps, window_size, window_size, num_channels//num_attention_heads)
+    """
+    batch_size, timesteps, height, width, num_channels = input_feature.shape
+    input_feature = input_feature.view(
+        batch_size,
+        timesteps,
+        height // window_size[0],
+        window_size[0],
+        width // window_size[1],
+        window_size[1],
+        num_attention_heads,
+        num_channels // num_attention_heads,
+    )
+    windows = input_feature.permute(0, 2, 4, 6, 1, 3, 5, 7).contiguous()
```

Comment:
:dizzy_face: wow, that's a hefty permute - maybe a comment here would help

Reply:
done :+1: